## Tag-Along Code for **Stage B Quiz**

In [21]:
!pip install gdown

In [22]:
import gdown

# Replace 'YOUR_FILE_ID' with the actual file ID from the shareable link.
file_id = '1Eru_UHVc3WLHVveC9Q8K9QUxlzYeHt18'

# Specify the destination file path and name.
output = 'energydata_complete.csv'  # You can change the file name and extension.

# Download the file.
gdown.download(f'https://drive.google.com/uc?id={"1Eru_UHVc3WLHVveC9Q8K9QUxlzYeHt18"}', output, quiet=False)


Downloading...
From: https://drive.google.com/uc?id=1Eru_UHVc3WLHVveC9Q8K9QUxlzYeHt18
To: /content/energydata_complete.csv
100%|██████████| 12.0M/12.0M [00:00<00:00, 132MB/s]


'energydata_complete.csv'

In [23]:
import pandas as pd

# Import the data and view the tranpose of the first 5 rows
df = pd.read_csv('energydata_complete.csv')  # Use the same filename you specified in the download step.
df.head().T

0                    1                    2  \
date         2016-01-11 17:00:00  2016-01-11 17:10:00  2016-01-11 17:20:00   
Appliances                    60                   60                   50   
lights                        30                   30                   30   
T1                         19.89                19.89                19.89   
RH_1                   47.596667            46.693333                 46.3   
T2                          19.2                 19.2                 19.2   
RH_2                       44.79              44.7225            44.626667   
T3                         19.79                19.79                19.79   
RH_3                       44.73                44.79            44.933333   
T4                          19.0                 19.0            18.926667   
RH_4                   45.566667              45.9925                45.89   
T5                     17.166667            17.166667            17.166667   
RH_5                        55.2                 55.2                55.09   
T6                      7.026667             6.833333                 6.56   
RH_6                   84.256667            84.063333            83.156667   
T7                          17.2                 17.2                 17.2   
RH_7                   41.626667                41.56            41.433333   
T8                          18.2                 18.2                 18.2   
RH_8                        48.9            48.863333                48.73   
T9                     17.033333            17.066667                 17.0   
RH_9                       45.53                45.56                 45.5   
T_out                        6.6             6.483333             6.366667   
Press_mm_hg                733.5                733.6                733.7   
RH_out                      92.0                 92.0                 92.0   
Windspeed                    7.0             6.666667             6.333333   
Visibility                  63.0            59.166667            55.333333   
Tdewpoint                    5.3                  5.2                  5.1   
rv1                    13.275433            18.606195            28.642668   
rv2                    13.275433            18.606195            28.642668   

                               3                    4  
date         2016-01-11 17:30:00  2016-01-11 17:40:00  
Appliances                    50                   60  
lights                        40                   40  
T1                         19.89                19.89  
RH_1                   46.066667            46.333333  
T2                          19.2                 19.2  
RH_2                       44.59                44.53  
T3                         19.79                19.79  
RH_3                        45.0                 45.0  
T4                         18.89                18.89  
RH_4                   45.723333                45.53  
T5                     17.166667                 17.2  
RH_5                       55.09                55.09  
T6                      6.433333             6.366667  
RH_6                   83.423333            84.893333  
T7                     17.133333                 17.2  
RH_7                       41.29                41.23  
T8                          18.1                 18.1  
RH_8                       48.59                48.59  
T9                          17.0                 17.0  
RH_9                        45.4                 45.4  
T_out                       6.25             6.133333  
Press_mm_hg                733.8                733.9  
RH_out                      92.0                 92.0  
Windspeed                    6.0             5.666667  
Visibility                  51.5            47.666667  
Tdewpoint                    5.0                  4.9  
rv1                    45.410389            10.084097  
rv2                    45.410389            10.084097

In [24]:
df.info()    # Get information about the dataset

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19735 entries, 0 to 19734
Data columns (total 29 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   date         19735 non-null  object 
 1   Appliances   19735 non-null  int64  
 2   lights       19735 non-null  int64  
 3   T1           19735 non-null  float64
 4   RH_1         19735 non-null  float64
 5   T2           19735 non-null  float64
 6   RH_2         19735 non-null  float64
 7   T3           19735 non-null  float64
 8   RH_3         19735 non-null  float64
 9   T4           19735 non-null  float64
 10  RH_4         19735 non-null  float64
 11  T5           19735 non-null  float64
 12  RH_5         19735 non-null  float64
 13  T6           19735 non-null  float64
 14  RH_6         19735 non-null  float64
 15  T7           19735 non-null  float64
 16  RH_7         19735 non-null  float64
 17  T8           19735 non-null  float64
 18  RH_8         19735 non-null  float64
 19  T9  

In [25]:
df.isnull().sum()   # Check for null values in dataset

date           0
Appliances     0
lights         0
T1             0
RH_1           0
T2             0
RH_2           0
T3             0
RH_3           0
T4             0
RH_4           0
T5             0
RH_5           0
T6             0
RH_6           0
T7             0
RH_7           0
T8             0
RH_8           0
T9             0
RH_9           0
T_out          0
Press_mm_hg    0
RH_out         0
Windspeed      0
Visibility     0
Tdewpoint      0
rv1            0
rv2            0
dtype: int64

Question 17: From the dataset, fit a linear model on the relationship between the temperature in the living room in Celsius (x = T2) and the temperature outside the building (y = T6). What is the Root Mean Squared error in three D.P?

In [26]:
# Import necessary libraries
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

# Prepare your data
X = df['T2'].values.reshape(-1, 1)
y = df['T6']

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Fit a linear regression model
model = LinearRegression()
model.fit(X_train, y_train)

# Make predictions
y_pred = model.predict(X_test)

# Calculate RMSE
rmse = np.sqrt(mean_squared_error(y_test, y_pred))

# Print the RMSE rounded to three decimal places
print(f"Root Mean Squared Error (RMSE): {rmse:.3f}")


Root Mean Squared Error (RMSE): 3.633


Remove the following columns: [“date”, “lights”]. The target variable is “Appliances”. Use a 70-30 train-test set split with a  random state of 42 (for reproducibility). Normalize the dataset using the MinMaxScaler (Hint: Use the MinMaxScaler fit_transform and transform methods on the train and test set respectively). Run a multiple linear regression using the training set. Answer the following questions:

Question 18: What is the Mean Absolute Error (in three decimal places) for the  training set?

In [27]:
from sklearn.preprocessing import MinMaxScaler
from sklearn.metrics import mean_absolute_error

# Remove the specified columns
data = df.drop(columns=["date", "lights"])

# Define the features (X) and the target variable (y)
X = data.drop(columns=["Appliances"])
y = data["Appliances"]

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# Initialize the MinMaxScaler and fit_transform on the training set
scaler = MinMaxScaler()
X_train_scaled = scaler.fit_transform(X_train)

# Transform the test set using the scaler fitted on the training set
X_test_scaled = scaler.transform(X_test)

# Initialize and fit the multiple linear regression model
model = LinearRegression()
model.fit(X_train_scaled, y_train)

# Make predictions on the training set
y_train_pred = model.predict(X_train_scaled)

# Calculate the Mean Absolute Error (MAE) on the training set
mae = mean_absolute_error(y_train, y_train_pred)

# Print the MAE rounded to three decimal places
print(f"Mean Absolute Error (MAE) for the training set: {mae:.3f}")


Mean Absolute Error (MAE) for the training set: 53.742


Question 19: What is the Root Mean Squared Error (in three decimal places) for the training set?

In [28]:

# Make predictions on the training set
y_train_pred = model.predict(X_train_scaled)

# Calculate the Root Mean Squared Error (RMSE) on the training set
rmse = np.sqrt(mean_squared_error(y_train, y_train_pred))

# Print the RMSE rounded to three decimal places
print(f"Root Mean Squared Error (RMSE) for the training set: {rmse:.3f}")


Root Mean Squared Error (RMSE) for the training set: 95.216


Question 20: What is the Mean Absolute Error (in three decimal places) for test set?

In [29]:

# Make predictions on the test set
y_test_pred = model.predict(X_test_scaled)

# Calculate the Mean Absolute Error (MAE) on the test set
mae = mean_absolute_error(y_test, y_test_pred)

# Print the MAE rounded to three decimal places
print(f"Mean Absolute Error (MAE) for the test set: {mae:.3f}")


Mean Absolute Error (MAE) for the test set: 53.643


Question 21: What is the Root Mean Squared Error (in three decimal places) for test set?

In [30]:

# Make predictions on the test set
y_test_pred = model.predict(X_test_scaled)

# Calculate the Root Mean Squared Error (RMSE) on the test set
rmse = np.sqrt(mean_squared_error(y_test, y_test_pred))

# Print the RMSE rounded to three decimal places
print(f"Root Mean Squared Error (RMSE) for the test set: {rmse:.3f}")


Root Mean Squared Error (RMSE) for the test set: 93.640


Question 22: Did the Model above overfit to the training set

No, based on the RMSE values, it does not seem that the model has overfit the training set. The closeness of the RMSE values on the training and test sets suggests that the model is performing consistently on both datasets.

Question 23: Train a ridge regression model with default parameters. Is there any change to the root mean squared error (RMSE) when evaluated on the test set?

In [34]:
from sklearn.linear_model import Ridge

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)
ridge_model = Ridge()  # Default alpha (regularization strength) is 1.0
ridge_model.fit(X_train, y_train)
y_pred = ridge_model.predict(X_test)
rmse = np.sqrt(mean_squared_error(y_test, y_pred))
print(f"Root Mean Squared Error (RMSE): {rmse:.3f}")


Root Mean Squared Error (RMSE): 93.641


RMSE for test set is 93.640 and RMSE for Ridge regression is 93.641. As seen, the difference in RMSE is insignificantly small which makes the RMSE values the same.

Question 24: Train a lasso regression model with default value and obtain the new feature weights with it. How many of the features have non-zero feature weights?

In [35]:
from sklearn.linear_model import Lasso
lasso_model = Lasso()  # Default alpha (regularization strength) is 1.0
lasso_model.fit(X, y)
feature_weights = lasso_model.coef_
non_zero_feature_count = np.sum(feature_weights != 0)
print("Number of features with non-zero weights:", non_zero_feature_count)


Number of features with non-zero weights: 19


Question 25: What is the new RMSE with the Lasso Regression on the test set?

In [20]:
y_pred = lasso_model.predict(X_test)
rmse = np.sqrt(mean_squared_error(y_test, y_pred))
print(f"Root Mean Squared Error (RMSE) with the Lasso Regression on the test set: {rmse:.3f}")

Root Mean Squared Error (RMSE) with the Lasso Regression on the test set: 93.750
